api/agent/agent.go

```diff
@@ -683,21 +685,60 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 			state.UpdateState(ctx, ContainerStateIdle, call.slots)
 			s := call.slots.queueSlot(&hotSlot{done, errC, container, nil})
 
-			select {
-			case <-s.trigger:
-			case <-time.After(time.Duration(call.IdleTimeout) * time.Second):
-				if call.slots.ejectSlot(s) {
-					logger.Info("Canceling inactive hot function")
-					return
-				}
-			case <-ctx.Done(): // container shutdown
-				if call.slots.ejectSlot(s) {
-					return
+			isFrozen := false
+			elapsed := time.Duration(0)
+			ticker := freezeTicker
+
+			for {
+				select {
+				case <-s.trigger: // slot already consumed
+				case <-ctx.Done(): // container shutdown
+				case <-a.shutdown: // server shutdown
+				case <-time.After(ticker):
+					elapsed += ticker
+					if elapsed < idleTimeout {
+
+						if ticker == freezeTicker {
+							if !isFrozen {
+								err := a.driver.Freeze(ctx, container)
+								if err != nil {
+									logger.WithError(err).Error("freeze error")
+									return
+								}
+								isFrozen = true
+							}
+							ticker = ejectTicker
+							continue
+						}
+
+						// if someone is waiting for resource in our slot queue, we must not terminate,
+						// otherwise, see if other slot queues have resource waiters.
+						stats := call.slots.getStats()
+						if stats.containerStates[ContainerStateWait] > 0 ||
+							a.resources.GetResourceTokenWaiterCount() <= 0 {
+							continue
+						}
+
+						logger.Debug("attempting hot function eject")
+					}
 				}
-			case <-a.shutdown: // server shutdown
-				if call.slots.ejectSlot(s) {
+				break
+			}
+
+			// if we can eject token, that means we are here due to
+			// abort/shutdown/timeout, attempt to eject and terminate,
+			// otherwise continue processing the request
+			if call.slots.ejectSlot(s) {
```

Comment:
this seems a little dicey, but did some thinking and couldn't dig up any apparent issues with doing this other than relying on ejectSlot's behavior.

Reply:
This is a bit tricky code here. I attempted to simplify the loop/flow, but maybe it's worse now. It assumes if we can eject a token, then we are here due to timeouts, shutdowns, etc. Otherwise, the I/O select should not land here without the consumer pulling the trigger. If consumer pulled the trigger, then we should not be able to eject. Costs an atomic cas, so it's fine wrt perf IMHO.

Reply:
yea, I couldn't think up any case, since if the slot is taken this will fail I think it covers most of the races. it seems correct